neuralmonkey/decoders/transformer.py

```diff
@@ -143,101 +151,122 @@ def embedded_train_inputs(self) -> tf.Tensor:
+            normalized_states = layer_norm(prev_layer.temporal_states)
+
             # TODO handle histories
+            # Run self-attention
             self_context, _ = attention(
-                queries=prev_layer.temporal_states,
-                keys=prev_layer.temporal_states,
-                values=prev_layer.temporal_states,
+                queries=normalized_states,
+                keys=normalized_states,
+                values=normalized_states,
                 keys_mask=prev_layer.temporal_mask,
                 num_heads=self.n_heads_self,
                 masked=True,
                 dropout_callback=lambda x: dropout(
                     x, self.attention_dropout_keep_prob, self.train_mode))
 
-            return dropout(
+            self_context = dropout(
                 self_context, self.dropout_keep_prob, self.train_mode)
 
-    def encoder_attention(self, level: int, queries: tf.Tensor) -> tf.Tensor:
+            # Residual connections
+            return self_context + prev_layer.temporal_states
+
+    def encoder_attention_sublayer(
+            self, queries: tf.Tensor) -> tf.Tensor:
 
-        with tf.variable_scope("dec_inter_att_level_{}".format(level),
+        with tf.variable_scope("encdec_attention",
                                reuse=tf.AUTO_REUSE):
             encoder_att_states = get_attention_states(self.encoder)
             encoder_att_mask = get_attention_mask(self.encoder)
 
+            # Layer normalization
+            normalized_queries = layer_norm(queries)
+
             # TODO handle histories
+            # Attend to the encoder
             encoder_context, _ = attention(
-                queries=queries,
+                queries=normalized_queries,
                 keys=encoder_att_states,
                 values=encoder_att_states,
                 keys_mask=encoder_att_mask,
                 num_heads=self.n_heads_enc,
                 dropout_callback=lambda x: dropout(
                     x, self.attention_dropout_keep_prob, self.train_mode))
 
-            return dropout(
+            encoder_context = dropout(
                 encoder_context, self.dropout_keep_prob, self.train_mode)
 
-    def layer(self, level: int, inputs: tf.Tensor,
-              mask: tf.Tensor) -> TransformerLayer:
-        # Recursive implementation. Outputs of the zeroth layer are the inputs
-        if level == 0:
-            norm_inputs = tf.contrib.layers.layer_norm(
-                inputs, begin_norm_axis=2)
-            return TransformerLayer(norm_inputs, mask)
+            # Residual connections
+            return encoder_context + queries
 
-        # Compute the outputs of the previous layer
-        prev_layer = self.layer(level - 1, inputs, mask)
+    def feedforward_sublayer(
+            self, layer_input: tf.Tensor) -> tf.Tensor:
 
-        # Run self-attention
-        self_context = self.masked_self_attention(level, prev_layer)
+        with tf.variable_scope("feedforward"):
```

Comment:
Proč tady neni `reuse=tf.AUTO_REUSE` ?

Reply:
Do tech *_sublayer metod lezeme jenom pri vytvareni daneho enkoderu/dekoderu a kazda sublayer je v ramci modelu zapojena prave jednou (coz je imho spravne). Vypnuty AUTO_REUSE je dobry test spravnosti kodu (pokud bychom tam lezli znovu v ramci stejneho scopu, tak chceme aby to spadlo, ne?).